overlord/snapstate/snapstate.go

```diff
@@ -1063,6 +1064,44 @@ func InstallWithDeviceContext(ctx context.Context, st *state.State, name string,
 	return doInstall(st, &snapst, snapsup, 0, fromChange, nil)
 }
 
+func InstallPathMany(st *state.State, sideInfos []*snap.SideInfo, paths, instanceNames, channels []string, flags Flags) ([]*state.TaskSet, []*snap.Info, error) {
+	tasksets := make([]*state.TaskSet, 0, len(paths))
+	infos := make([]*snap.Info, 0, len(paths))
+
+	// transform paths into absolutes so we can try to deduplicate them
+	for i, path := range paths {
+		newPath, err := filepath.Abs(path)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot make %q into absolute path: %w", path, err)
+		}
+		paths[i] = newPath
+	}
+	paths = strutil.Deduplicate(paths)
+
+	for i, si := range sideInfos {
+		path := paths[i]
+		var instanceName string
+		if instanceNames != nil {
+			instanceName = instanceNames[i]
+		}
+		var channel string
+		if channels != nil {
+			channel = channels[i]
+		}
+
+		ts, info, err := InstallPath(st, si, path, instanceName, channel, flags)
```

Comment:
as I tried to explain the goal is really to share logic with UpdateMany because we need the sorting done there to be able to leverage the work Maciej is doing and also our main use cases are really local install that are updates, so we cannot just reuse InstallPath like this, instead we need to reuse the later parts of updateManyFiltered and take setups bits derived from Installmany and InstallPath